Tests/DataPipeline/DataPipelineCompatibilityTests.swift

```diff
@@ -0,0 +1,438 @@
+        XCTAssertEqual(filteredEvents.count, 1, "too many events received")
+
+        guard let savedEvent = filteredEvents.last else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertEqual(savedEvent[keyPath: "userId"] as? String, givenIdentifier)
+        XCTAssertMatches(
+            savedEvent[mapKeyPath: "traits"],
+            givenBody,
+            withTypeMap: [["age"]: Int.self]
+        )
+    }
+
+    func test_eventBeforeIdentify_expectFinalJSONHasNoUserId() {
+        let givenEvent = String.random
+
+        customerIO.track(name: givenEvent)
+
+        let allEvents = readTypeFromStorage(key: Storage.Constants.events)
+        let filteredEvents = allEvents.filter { $0.eventType == "track" }
+        XCTAssertEqual(filteredEvents.count, 1, "too many events received")
+
+        guard let savedEvent = filteredEvents.last else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertNil(savedEvent[keyPath: "userId"] as? String)
+    }
+
+    func test_eventAfterIdentify_expectFinalJSONHasCorrectUserId() {
+        let givenIdentifier = String.random
+        let givenEvent = String.random
+
+        customerIO.identify(userId: givenIdentifier)
+        customerIO.track(name: givenEvent)
+
+        let allEvents = readTypeFromStorage(key: Storage.Constants.events)
+        let filteredEvents = allEvents.filter { $0.eventType == "track" }
+        XCTAssertEqual(filteredEvents.count, 1, "too many events received")
+
+        guard let savedEvent = filteredEvents.last else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertEqual(savedEvent[keyPath: "userId"] as? String, givenIdentifier)
+    }
+
+    // MARK: device
+
+    func test_registerDeviceToken_expectFinalJSONHasCorrectKeysAndValues() {
+        let givenIdentifier = String.random
+        let givenToken = String.random
+        let expectedData = deviceInfoMock.configureWithMockData()
+
+        customerIO.identify(identifier: givenIdentifier)
+        customerIO.registerDeviceToken(givenToken)
+
+        let allEvents = readTypeFromStorage(key: Storage.Constants.events)
+        let filteredEvents = allEvents.filter { $0.eventType == "track" }
+        XCTAssertEqual(filteredEvents.count, 1, "too many events received")
+
+        guard let savedEvent = filteredEvents.first else {
+            XCTFail("saved event must not be nil")
+            return
+        }
+
+        XCTAssertEqual(savedEvent[keyPath: "event"] as? String, "Device Created or Updated")
+        XCTAssertEqual(savedEvent[keyPath: "userId"] as? String, givenIdentifier)
+        XCTAssertEqual(savedEvent[keyPath: "context.device.token"] as? String, givenToken)
+        XCTAssertFalse(savedEvent.containsKey("last_used"))
+        XCTAssertFalse(savedEvent.containsKey("platform"))
```

Comment:
are these all the required attributes? if not i think we should add all the required attributes to add a new device?

Reply:
Added `type` as well. If you have more mind, please share so I can add them too